Libraries/Opc.Ua.Server/NodeManager/MasterNodeManager.cs

```diff
@@ -468,25 +468,32 @@ public virtual object GetManagerHandle(NodeId nodeId, out INodeManager nodeManag
 
             // use the namespace index to select the node manager.
             int index = nodeId.NamespaceIndex;
+            INodeManager[] nodeManagers = null;
 
             lock (m_namespaceManagers.SyncRoot)
             {
                 // check if node managers are registered - use the core node manager if unknown.
-                if (index >= m_namespaceManagers.Length || m_namespaceManagers[index] == null)
+                if (index < m_namespaceManagers.Length && m_namespaceManagers[index] != null)
                 {
-                    handle = m_nodeManagers[1].GetManagerHandle(nodeId);
+                    nodeManagers = m_namespaceManagers[index];
+                }
+            }
 
```

Comment:
the lock should remain around all loops

Reply:
I agree it would generally safeguard, but this is a deadlock waiting to happen (and actually happened in #1851) since CustomNodeManager2.GetManagerHandle locks its public Lock, which besides being used also in all derived node managers, its also used by MonitoredNode and ServerInternalData.
Since m_nodeManagers is modified during startup and no other thread apparently operates on the List, I think it is safe to unwrap it from the Lock(m_namespaceManagers.SyncRoot) as the code stands now.
I see no other timely feasible workaround from this deadlock state (#1851).

Reply:
better use ReaderWriterLockSlim instead of Lock(m_namespaceManagers.SyncRoot) , there are rare cases when the nodemanager array changes, so most of the time threads can start to get the manager handle concurrently.